### ISSUE TITLE
How do I exclude my custom lock screen layer when sharing a screen?

### ISSUE THREAD

**Issue by zoujiaqing** [livekit/client-sdk-android#561]
I have a mobile workstation.
I want to see it run and operate remotely.
But my app has a lock screen that hides the phone's screen.
How do I filter out this mask when I look at it remotely?

livekit sdk has an API same this code?

```
val localVideoTrackOptions = LocalVideoTrackOptions.Builder()
    .excludeApps(listOf("com.example.customlock"))  // exclude this
    .build()
```

**Comment by davidliu** [comment]
I'm not sure what you want to do. Do you want to exclude an app from being shared from the screen share?

If it helps, you can write your response in Chinese, I can understand a bit.

**Comment by zoujiaqing** [comment]
好的兄弟，我说一下我的思路。

我在做的是手机群控系统，但是手机托管在第三方的机房中，我需要这一个自定义的app来锁定这个屏幕，让第三方机房无法看到我手机的界面，但是我在屏幕共享的时候没有排除这个锁屏界面的显示，我之前的方案是要修改 framework层达到目的，但是毕竟手机系统不是我们的，所以 VisualDisplay 加上修改底层 framework 的方案不行。按理说通过 VisualDisplay 可以让虚拟显示器不显示这个自定义锁屏activity，然后推流这个虚拟显示器的内容即可达到需要的效果。但是 livekitsdk for android 将 visualdisplay 封装在内部了吧，不知道我要的这个方式如何能实现。

谢谢你。

**Comment by zoujiaqing** [comment]
Do you know my mean? @davidliu

**Comment by zoujiaqing** [comment]
Yes, I need create a custom VirtualDiaplay object transmitted to room.localParticipant.publishVideoTrack(track)?

@davidzhao

**Comment by davidliu** [comment]
Yes. You can reference the code I linked above to make your own implementation.

**Comment by zoujiaqing** [comment]
I created a custom VideoCapturer. But what the remote video is seeing is not what my video is about. How to make this video replace the screen recorded video stream?

My code:

```kotlin

    private fun startScreenCapture(intentData: Intent) {
        val localParticipant = room.localParticipant

        GlobalScope.launch {
            val resolution = VideoResolution.H240
            val frameRate = 6
            val captureParams = getCaptureParams(context, resolution, frameRate)

            val excludedPackages = setOf("com.xxx.remote")

            val screenShareVideoCapturer = ScreenShareVideoCapturer(context, intentData, captureParams, excludedPackages)

            val track = localParticipant.createVideoTrack(
                capturer = screenShareVideoCapturer,
                name = "screenshare"
            )

            localParticipant.publishVideoTrack(
                track = track,
                options = VideoTrackPublishOptions(
                    simulcast = false,
                    videoEncoding = createCustomVideoEncoding(context, resolution, frameRate)
                )
            )

            screenShareVideoCapturer.startCapture(captureParams.width, captureParams.height, frameRate)
        }
    }
```

**Comment by davidliu** [comment]
Your custom video capturer needs to pass the frames to the `capturerObserver` that's given in the `initialize` method.

**Comment by zoujiaqing** [comment]
Yes, it's can capture video, now!
thanks ;)

but, I want shared screen.

How to create ScreencastTrack?

**Comment by zoujiaqing** [comment]
My custom VideoCapturer can capture android screen video publish to livekit server.

but, the video have big black background. why?

I use room.localParticipant.setScreenShareEnabled(true, intentData) not black background.

**Comment by davidliu** [comment]
> Yes, it's can capture video, now! thanks ;)
> 
> but, I want shared screen.
> 
> How to create ScreencastTrack?

It doesn't need to be a `LocalScreencastVideoTrack`, that's for our internal handling of it for user convenience, but you can just publish it as a `LocalVideoTrack`.
 
> Because the receiver is using the interface for watch screen sharing.

If you mean the receiver is getting the track through `Track.Source.SCREEN_SHARE`, you can choose which source you want to publish it as:

```
room.localParticipant.publishVideoTrack(
    track = yourTrack,
    options = VideoTrackPublishOptions(
        base = room.screenShareTrackPublishDefaults,
        source = Track.Source.SCREEN_SHARE
    )
)
```

> but, the video have big black background. why?

You may be setting the wrong dimensions when capturing the screen share.
